clang/lib/Sema/SemaInit.cpp

```diff
@@ -7575,15 +7577,27 @@ static void visitLifetimeBoundArguments(IndirectLocalPath &Path, Expr *Call,
     Path.pop_back();
   };
 
-  if (ObjectArg && implicitObjectParamIsLifetimeBound(Callee))
-    VisitLifetimeBoundArg(Callee, ObjectArg);
-
   bool CheckCoroCall = false;
   if (const auto *RD = Callee->getReturnType()->getAsRecordDecl()) {
     CheckCoroCall = RD->hasAttr<CoroLifetimeBoundAttr>() &&
                     RD->hasAttr<CoroReturnTypeAttr>() &&
                     !Callee->hasAttr<CoroDisableLifetimeBoundAttr>();
   }
+
+  if (ObjectArg) {
+    bool CheckCoroObjArg = CheckCoroCall;
+    // Ignore `__promise.get_return_object()` as it is not lifetimebound.
+    if (CheckCoroObjArg && Callee->getDeclName().isIdentifier() &&
+        Callee->getName() == "get_return_object")
+      CheckCoroObjArg = false;
```

Comment:
The hardcoded name are not good. What's wrong with it? And if necessary, we'd better to check it with the built `CoroutineBodyStmt::getReturnStmt` (and maybe `CoroutineBodyStmt::getReturnStmtOnAllocFailure`?) If that is not built this time, can we perform that in `Sema::CheckCompletedCoroutineBody`? Maybe that will require a little bit refactor work, but I guess it may be worthy.

Reply:
There is also a similar use of the name in `CheckCoroutineWrapper`, I suggest to unify the check for matching this particular function across the two call sites.

The motivation to disable this for `get_return_object` is clear: it's often used as an implementation detail, so warning there creates noise. There could be better ways to do it, though.

Reply:
I have refactored the check. The check still uses hardcoded names though.

Reply:
hmmm it looks worse now... if our previous suggestion is hard to implement, I'll suggest to do a coroutine lookup here. For example, we know a class is marked with coro_lifetimebound attribute, then we should be able to find the corresponding promise_type, right? And if yes, we can give exceptions to the member functions of the promise_type so that we can cover the case here.

Reply:
Promise type lookup would need to go through coroutine traits. We could lookup `coroutine_traits<R>::promise_type` but does not work with `coroutine_traits` specialized for non-empty parameter list.
The new approach implicitly adds `[[clang::coro_wrapper]]` and `[[clang::coro_disable_lifetimebound]]` to `get_return_type` found through a promise lookup.
WDYT ?